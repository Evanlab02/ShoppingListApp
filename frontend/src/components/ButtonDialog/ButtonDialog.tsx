// Third party imports
import { SlDialog, SlButton } from '@shoelace-style/shoelace/dist/react';

// File imports
import { ButtonDialogProps } from './helpers/interfaces';

/**
 * Button dialog component.
 * 
 * @param props Button dialog props
 * @returns Button dialog component
 */
export default function ButtonDialog(props: ButtonDialogProps) {
    const { onClose, open, title, text = "", buttonValues } = props;

    return (
        <>
            <SlDialog label={title} open={open} onSlAfterHide={onClose}>
                <p>{text}</p>
                <SlButton slot="footer" variant="danger" onClick={onClose}>
                    Close
                </SlButton>
                <>
                    {buttonValues.map((buttonValue, index) => {
                        return (
                            <SlButton
                                key={index}
                                slot="footer"
                                variant={buttonValue.variant}
                                onClick={buttonValue.onClick}
                            >
                                {buttonValue.text}
                            </SlButton>
                        );
                    })}
                </>
            </SlDialog >
        </>
    );
}
