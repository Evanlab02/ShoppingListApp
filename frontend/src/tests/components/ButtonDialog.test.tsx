import { beforeEach, expect, test, vi } from "vitest";
import { fireEvent, render } from "@testing-library/react";
import ButtonDialog from "../../components/ButtonDialog/ButtonDialog";

beforeEach(() => {
	document.getElementsByTagName("html")[0].innerHTML = "";
});

test("Button Dialog renders and functions correctly.", async () => {
	const fn = vi.fn();
    const button1Click = vi.fn();
    const button2Click = vi.fn();

    const { findByTestId } = render(
        <div data-testid="test-button-dialog">
            <ButtonDialog 
                open={true}
                onClose={fn}
                title="Testing Button Dialog"
                buttonValues={[
                    {
                        variant: "default",
                        text: "Hi",
                        onClick: button1Click
                    },
                    {
                        variant: "danger",
                        text: "Goodbye",
                        onClick: button2Click
                    }
                ]}
            />
        </div>
	);

	const btnDialog = await findByTestId("test-button-dialog");
	expect(btnDialog).toMatchSnapshot();

    
    const closeButton = await findByTestId("sa-btn-dl-close");
    fireEvent.click(closeButton);
    expect(fn).toBeCalledTimes(1);

    const customButton1 = await findByTestId("sa-btn-dl-custom-0");
    fireEvent.click(customButton1);
    expect(button1Click).toBeCalledTimes(1);

    const customButton2 = await findByTestId("sa-btn-dl-custom-1");
    fireEvent.click(customButton2);
    expect(button2Click).toBeCalledTimes(1);
});