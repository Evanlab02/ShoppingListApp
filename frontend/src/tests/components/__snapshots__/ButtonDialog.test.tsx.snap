// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Dialog renders and functions correctly. 1`] = `
<div
  data-testid="test-button-dialog"
>
  <sl-dialog
    label=""
  >
    <p />
    <sl-button
      data-optional=""
      data-testid="sa-btn-dl-close"
      data-valid=""
      size="medium"
      variant="default"
    >
      Close
    </sl-button>
    <sl-button
      data-optional=""
      data-testid="sa-btn-dl-custom-0"
      data-valid=""
      size="medium"
      variant="default"
    >
      Hi
    </sl-button>
    <sl-button
      data-optional=""
      data-testid="sa-btn-dl-custom-1"
      data-valid=""
      size="medium"
      variant="default"
    >
      Goodbye
    </sl-button>
  </sl-dialog>
</div>
`;
